src/cirq_iqm/devices/apollo.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2020–2022 Cirq on IQM developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+IQM's Apollo quantum architecture.
+"""
+from __future__ import annotations
+
+from math import pi as PI
+from typing import Optional
+
+import cirq
+from cirq import ops
+
+from .iqm_device import IQMDevice
+
+PI_2 = PI / 2
+
+# common gates used in gate decompositions
+CZ = ops.CZPowGate()
+Lx = ops.rx(PI_2)
+Lxi = ops.rx(-PI_2)
+Ly = ops.ry(PI_2)
+Lyi = ops.ry(-PI_2)
+
+
+class Apollo(IQMDevice):
+    """IQM's twenty-qubit transmon device.
+
+    The qubits are connected thus::
+
+            QB20   QB17
+            /  |   /  |
+         QB19  QB16  QB12
+        /  |   /  |  /  |
+     QB18  QB15  QB11  QB7
+        |  /  |  /  |  /
+        QB14  QB10  QB6
+        /  |  /  |  /
+     QB13  QB9    QB5
+        |  /  |  /  |
+        QB8   QB4   QB2
+          |  /  |  /
+          QB3   QB1
+
+    where the lines denote which qubit pairs can be subject to two-qubit gates.
+
+    Each qubit can be rotated about any axis in the xy plane by an arbitrary angle.
+    Apollo thus has native PhasedXPowGate, XPowGate, and YPowGate gates. The two-qubit gate CZ is
+    native, as well. The qubits can be measured simultaneously or separately once, at the end of
+    the circuit.
+    """
+
+    QUBIT_COUNT = 20
+
+    CONNECTIVITY = (
+        {1, 2}, {1, 4},
+        {2, 5},
+        {3, 4}, {3, 8},
+        {4, 5}, {4, 9},
+        {5, 6}, {5, 10},
+        {6, 7}, {6, 11},
+        {7, 12},
+        {8, 9}, {8, 13},
+        {9, 10}, {9, 14},
+        {10, 11}, {10, 15},
+        {11, 12}, {11, 16},
+        {12, 17},
+        {13, 14},
+        {14, 15}, {14, 18},
+        {15, 16}, {15, 19},
+        {16, 17}, {16, 20},
+        {18, 19},
+        {19, 20}
+    )
+
+    NATIVE_GATES = (
+        ops.PhasedXPowGate,
+        ops.XPowGate,
+        ops.YPowGate,
+        ops.MeasurementGate
+    )
+
+    NATIVE_GATE_INSTANCES = (
+        ops.CZPowGate(),
+    )
+
+    def operation_decomposer(self, op: cirq.Operation) -> Optional[list[cirq.Operation]]:
```

Comment:
Should we move this function to the parent class instead of copying in various device implementations?

Reply:
I guess. But then we'll have to choose between:

1. `IQMDevice` stays as generic as now. `operation_decomposer` moves there, and becomes parameterized to allow for decomposition of different architectures. 
3. `IQMDevice` stays as generic as now. We introduce subclasses for compatible architectures (i.e. single subclass for Adonis&Apollo, since native gates are the same). `operation_decomposer` moves to that subclass.
4. `IQMDevice` becomes less generic and implements `operation_decomposer`, which somehow works for current Adonis&Apollo architecture, and is flexible for future different architectures.

Also, is it possible that two chips have the same native gates, but their decompositions still differ in certain details (e.g. global shift)?

Reply:
I think
1. is not great, since will quickly become cumbersome.
2. is something I had in mind as well, but probably does not make sense to implement now, maybe in the future?
3. is actually what I was thinking: a default operation decomposer in `IQMDevice`, and children can overwrite it as needed.
4. another though I had was to reconsider if we still need this functionality. Why is Cirq's default decomposition enough?

Decomposition of individual gates into the same native gate set should always be the same regardless of device. Circuit optimization routines could be different for different devices and end up in different circuits depending on device, but individual gate decomposition is always the same.

Reply:
@hay-k regarding 4: seems like Cirq default decomposition can handle `ISwapPowGate`. But that's not the case for `CZPowGate`, `ZZPowGate` and `ZPowGate` in our tests. Removing them from `operation_decomposer` break the tests.

So I propose for now to proceed with 3, i.e. moving `operation_decomposer` to `IQMDevice` while removing `ISwapPowGate` decomposition. Then investigate the possibility of getting rid of `operation_decomposer` in a separate issue.

@smite, we might need your feedback regarding this.

Reply:
Proceeded with 3: `operation_decomposer` is now implemented in `IQMDevice`, and children devices can override it (see Valkmusa device as an example of such overriding). 

The possibility of getting rid of `operation_decomposer` altogether will be discussed in a separate issue.